File: 2.2/zig/src/main.zig
const std = @import("std");
const mem = std.mem;
const allocator = std.debug.global_allocator;
const os = std.os;
const io = std.io;
const Buffer = std.Buffer;

pub fn main() anyerror!void {
    const inputFilePath = try os.path.join(allocator, "..", "input.txt");
    const rawBoxIds = try io.readFileAlloc(allocator, inputFilePath);

    var commonBoxIdBuffer = Buffer.initNull(allocator);
    defer commonBoxIdBuffer.deinit();

    var boxIdIter = mem.split(rawBoxIds, "\n");
    outer: while (boxIdIter.next()) |boxId| {
        var otherBoxIdIter = mem.split(rawBoxIds, "\n");
        while (otherBoxIdIter.next()) |otherBoxId| {
            try commonBoxIdBuffer.resize(0);
            var diffCount: u8 = 0;
            for (boxId) |char, i| {
                if (boxId[i] != otherBoxId[i]) {
                    diffCount += 1;
                } else {
                    try commonBoxIdBuffer.appendByte(char);
                }
            }
            
            if (diffCount == 1) {
                break :outer;
            }
        }
    }

    std.debug.warn("Common Box ID: {}", commonBoxIdBuffer.toOwnedSlice());
}

File: 2.2/rust/src/main.rs
use std::fs::File;
use std::string::String;
use std::io::prelude::*;
use std::char;

fn main() {
    let mut box_ids = String::new();
    File::open("./../input.txt")
        .expect("Unable to read from input.txt")
        .read_to_string(&mut box_ids)
        .expect("Unable to feed contents to string");

    let mut common_id: String = String::from("");
    'outer: for box_id in box_ids.lines() {
        for other_box_id in box_ids.lines() {
            let mut diff_count = 0;
            common_id.clear();
            
            for (first, second) in box_id.trim().chars().zip(other_box_id.trim().chars()) {
                if first == second {
                    common_id.push(first);
                } else {
                    diff_count += 1;
                }
            }

            if diff_count == 1 {
                break 'outer;
            }
        }
    }

    println!("Common Box ID: {}", common_id);
}
